File: src/main/java/gargoyle/ct/resource/internal/TempLocalResource.java
package gargoyle.ct.resource.internal;

import gargoyle.ct.resource.Resource;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;

public class TempLocalResource extends ClasspathResource {
    private TempLocalResource(ClassLoader loader, Resource base, String location) {
        super(loader, base, location);
    }

    public TempLocalResource(String location) {
        super(location);
    }

    @NotNull
    @Override
    protected TempLocalResource createResource(@Nullable Resource base, String location) {
        ClassLoader loader = getLoader(base);
        Resource resource = base == null ?
                new ClasspathResource(location) : new ClasspathResource(loader, base, location);
        if (!resource.exists()) {
            return new TempLocalResource(location);
        }
        try (InputStream stream = resource.getInputStream()) {
            File tempFile = File.createTempFile(getBaseName(), DOT + getExtension());
            Files.copy(stream, tempFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
            tempFile.deleteOnExit();
            return new TempLocalResource(loader, resource, tempFile.getPath());
        } catch (IOException e) {
            throw new IllegalArgumentException(location, e);
        }
    }

    @NotNull
    @Override
    public URL toURL() throws IOException {
        return new File(getLocation()).toURI().toURL();
    }

    @Override
    public boolean exists() {
        try {
            return exists(toURL());
        } catch (IOException ex) {
            return false;
        }
    }

    @NotNull
    public File toFile() throws IOException {
        return new File(toURL().getPath().substring(1));
    }
}

File: src/main/kotlin/gargoyle/ct/resource/internal/TempLocalResource.kt
package gargoyle.ct.resource.internal

import java.io.File
import java.io.IOException
import java.net.URL
import java.nio.file.Files
import java.nio.file.StandardCopyOption

class TempLocalResource : ClasspathResource {
    private constructor(loader: ClassLoader, base: ClasspathResource, location: String) : super(
        loader,
        base,
        location
    )

    constructor(location: String) : super(location)

    override fun createResource(
        base: VirtualResource<ClasspathResource>?,
        location: String
    ): VirtualResource<ClasspathResource> {
        val loader = getLoader(base)
        val resource: ClasspathResource =
            base?.let { ClasspathResource(loader, it as ClasspathResource, location) } ?: ClasspathResource(location)
        if (!resource.exists()) {
            return TempLocalResource(location)
        }
        try {
            resource.inputStream.use { stream ->
                val tempFile = File.createTempFile(baseName, DOT + extension)
                Files.copy(stream, tempFile.toPath(), StandardCopyOption.REPLACE_EXISTING)
                tempFile.deleteOnExit()
                return TempLocalResource(loader, resource, tempFile.path)
            }
        } catch (e: IOException) {
            throw IllegalArgumentException(location, e)
        }
    }

    @Throws(IOException::class)
    override fun toURL(): URL {
        return File(location).toURI().toURL()
    }

    override fun exists(): Boolean {
        return try {
            exists(toURL())
        } catch (ex: IOException) {
            false
        }
    }

    @Throws(IOException::class)
    fun toFile(): File {
        return File(toURL().path.substring(1))
    }
}
